src/Nethermind/Nethermind.State/PartialStorageProviderBase.cs

```diff
@@ -0,0 +1,230 @@
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+
+using System;
+using System.Collections.Generic;
+using Nethermind.Core;
+using Nethermind.Core.Collections;
+using Nethermind.Core.Crypto;
+using Nethermind.Core.Extensions;
+using Nethermind.Core.Resettables;
+using Nethermind.Logging;
+using Nethermind.Trie.Pruning;
+
+namespace Nethermind.State
+{
+    public abstract class PartialStorageProviderBase
+    {
+        protected readonly ResettableDictionary<StorageCell, StackList<int>> _intraBlockCache = new();
+
+        /// <summary>
+        /// EIP-1283
+        /// </summary>
+        protected readonly ResettableDictionary<StorageCell, byte[]> _originalValues = new();
+
+        protected readonly ResettableHashSet<StorageCell> _committedThisRound = new();
+
+        protected readonly ILogger _logger;
+        protected readonly ILogManager _logManager;
+
+        protected readonly ResettableDictionary<Address, StorageTree> _storages = new();
+
+        protected const int StartCapacity = Resettable.StartCapacity;
+        protected int _capacity = StartCapacity;
+        protected Change?[] _changes = new Change[StartCapacity];
+        protected int _currentPosition = Resettable.EmptyPosition;
+
+        // stack of snapshot indexes on changes for start of each transaction
+        // this is needed for OriginalValues for new transactions
+        protected readonly Stack<int> _transactionChangesSnapshots = new();
+
+        public PartialStorageProviderBase(ILogManager? logManager)
+        {
+            _logManager = logManager ?? throw new ArgumentNullException(nameof(logManager));
+            _logger = logManager.GetClassLogger<PartialStorageProviderBase>() ?? throw new ArgumentNullException(nameof(logManager));
+        }
+
+        public byte[] Get(StorageCell storageCell)
+        {
+            return GetCurrentValue(storageCell);
+        }
+
+        public void Set(StorageCell storageCell, byte[] newValue)
+        {
+            PushUpdate(storageCell, newValue);
+        }
+
+        public int TakeSnapshot(bool newTransactionStart)
+        {
+            if (_logger.IsTrace) _logger.Trace($"Storage snapshot {_currentPosition}");
+            if (newTransactionStart && _currentPosition != Resettable.EmptyPosition)
+            {
+                _transactionChangesSnapshots.Push(_currentPosition);
+            }
+
+            return _currentPosition;
+        }
+
+        public void Restore(int snapshot)
+        {
+            if (_logger.IsTrace) _logger.Trace($"Restoring storage snapshot {snapshot}");
+
+            if (snapshot > _currentPosition)
+            {
+                throw new InvalidOperationException($"{nameof(PartialStorageProviderBase)} tried to restore snapshot {snapshot} beyond current position {_currentPosition}");
```

Comment:
Nit: consider adding this nameof pattern to all the trace logs here too, else it will be hard to determine which apply to transient storage vs persistent storage